bl/drawable/inc/drawable.hpp

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include <list>
+#include <memory>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+class Drawable {
+   private:
+    template <typename T>
+    struct is_smart_ptr : std::false_type {};
+    template <typename T>
+    struct is_smart_ptr<std::shared_ptr<T>> : std::true_type {};
+    template <typename T>
+    struct is_smart_ptr<std::weak_ptr<T>> : std::true_type {};
+
+    // cant use unique_ptr in this case
+    //  template <typename T>
+    //  struct is_smart_ptr<std::unique_ptr<T>> : std::true_type {};
+
+   public:
+    // TODO change to shared_ptr only???
+    template <typename T, std::enable_if_t<is_smart_ptr<T>::value, bool> = true,
+              std::enable_if_t<
+                  std::is_base_of<Drawable, std::decay_t<decltype(*std::declval<T>())>>::value,
+                  bool> = true>
+    struct DrawablePosition {
+        int x;
+        int y;
+        T objHdl;
+    };
+
+    virtual ~Drawable() = default;
+    virtual char getColor() const = 0;
```

Comment:
How about creating seperate class/struct for pixel type, it could hold rgb values, if ever decide to use it with something which can support colors your code will support it.  And for the drowing angine whcih does not support colors it can any way take pixels as a data, but convert it to chars by it self.

Reply:
fixed